src/client/Client.h

```diff
@@ -1237,6 +1237,178 @@ class Client : public Dispatcher, public md_config_obs_t {
+      CRF->finish_io(r);
+    }
+
+    // For _read_async, we may not finish in one go, so be prepared for multiple
+    // calls to complete. All the handling though is in C_Read_Finisher.
+    void complete(int r) override {
+      finish(r);
+      if (CRF->iofinished)
+        delete this;
+    }
+  };
+
+  class C_Read_Sync_NonBlocking : public Context {
+  // When operating in non-blocking mode, what used to be done by _read_sync
+  // still needs to be handled, but it needs to be handled without blocking
+  // while still following the semantics. Note that _read_sync is actually
+  // asynchronous. it just uses condition variables to wait. Now instead, we use
+  // this Context class to synchronize the steps.
+  //
+  // The steps will be accomplished by complete/finish being called to complete
+  // each step, with complete only releasing this object once all is finally
+  // complete.
+  public:
+    C_Read_Sync_NonBlocking(Client *clnt, Context *onfinish, Fh *f, Inode *in,
+                            uint64_t fpos, uint64_t off, uint64_t len,
+                            bufferlist *bl, Filer *filer, int have_caps)
+      : clnt(clnt), onfinish(onfinish), f(f), in(in), off(off), len(len), bl(bl),
+        filer(filer), have_caps(have_caps)
+    {
+      left = len;
+      wanted = len;
+      read = 0;
+      pos = off;
+      fini = false;
+    }
+
+    void retry();
+
+  private:
+    Client *clnt;
+    Context *onfinish;
+    Fh *f;
+    Inode *in;
+    uint64_t off;
+    uint64_t len;
+    int left;
+    int wanted;
+    bufferlist *bl;
+    bufferlist tbl;
+    Filer *filer;
+    int have_caps;
+    int read;
+    uint64_t pos;
+    bool fini;
+
+    void finish(int r) override;
+
+    void complete(int r) override
+    {
+      finish(r);
+      if (fini)
+        delete this;
+    }
+  };
+
+  class C_Read_Async_Finisher : public Context {
+  public:
+    C_Read_Async_Finisher(Client *clnt, Context *onfinish, Fh *f, Inode *in,
+                          uint64_t fpos, uint64_t off, uint64_t len)
+      : clnt(clnt), onfinish(onfinish), f(f), in(in), off(off), len(len) {}
+
+  private:
+    Client *clnt;
+    Context *onfinish;
+    Fh *f;
+    Inode *in;
+    uint64_t off;
+    uint64_t len;
+
+    void finish(int r) override;
+  };
+
+  class C_Write_Finisher : public Context {
+  public:
+    void finish_io(int r);
+    void finish_onuninline(int r);
+
+    C_Write_Finisher(Client *clnt, Context *onfinish, bool dont_need_uninline,
+                     bool is_file_write, utime_t start, Fh *f, Inode *in,
+                     uint64_t fpos, int64_t offset, uint64_t size)
+      : clnt(clnt), onfinish(onfinish),
+        is_file_write(is_file_write), start(start), f(f), in(in), fpos(fpos),
+        offset(offset), size(size) {
+      iofinished_r = 0;
+      onuninlinefinished_r = 0;
+      iofinished = false;
+      onuninlinefinished = dont_need_uninline;
+    }
+
+    void finish(int r) override {
```

Comment:
Do we have people calling finish directly instead of complete? (And can we be sure nobody calls complete to avoid a double-free?)

Reply:
There are some mixups between calling complete and calling finish... I'm not absolutely sure I've got all those right.

Reply:
I will take one more look, but I think I can remove those finish overrides.